src/ImageClassifier/index.js

```diff
@@ -1,74 +1,257 @@
 // https://opensource.org/licenses/MIT
 
 /*
-Image classifier class based on ImageNet trained models.
-SqueezeNet and MobileNet supported.
+Image classifier class
 */
+/* eslint max-len: ["error", { "code": 180 }] */
 
-import { fromPixels, image } from 'deeplearn';
-import { SqueezeNet } from 'deeplearn-squeezenet';
-import { MobileNet } from './MobileNet';
-import { processVideo } from '../utils/imageUtilities';
+import * as tf from '@tensorflow/tfjs';
+import { IMAGENET_CLASSES } from './../utils/IMAGENET_CLASSES';
+import { processVideo, imgToTensor } from '../utils/imageUtilities';
+
+const DEFAULTS = {
+  learningRate: 0.0001,
+  hiddenUnits: 100,
+  epochs: 20,
+  numClasses: 2,
+  batchSize: 0.4,
+};
 
 class ImageClassifier {
-  constructor(model) {
-    this.model = model;
-    this.readyPromise = null;
-    if (this.model === 'SqueezeNet') {
-      this.net = new SqueezeNet();
-    } else if (this.model === 'MobileNet') {
-      this.net = new MobileNet();
-    } else {
-      console.warn(`${model} is not a valid model. Using MobileNet as default.`);
-      this.net = new MobileNet();
-    }
+  constructor(video, options = {}, callback = () => {}) {
+    this.mobilenet = null;
+    this.imageSize = 224;
+    this.modelPath = 'https://storage.googleapis.com/tfjs-models/tfjs/mobilenet_v1_0.25_224/model.json';
+    this.topKPredictions = 10;
+    this.modelLoaded = false;
     this.video = null;
+    this.waitingPredictions = [];
+
+    // Props for retraining mobilenet
+    this.hasAnyTrainedClass = false;
+    this.customModel = null;
+    this.epochs = options.epochs || DEFAULTS.epochs;
+    this.hiddenUnits = options.hiddenUnits || DEFAULTS.hiddenUnits;
+    this.numClasses = options.numClasses || DEFAULTS.numClasses;
+    this.learningRate = options.learningRate || DEFAULTS.learningRate;
+    this.batchSize = options.batchSize || DEFAULTS.batchSize;
+    this.isPredicting = false;
+    this.mapStringToIndex = [];
+
+    if (video instanceof HTMLVideoElement) {
+      this.video = processVideo(video, this.imageSize);
+    }
+
+    this.loadModel().then((net) => {
+      this.modelLoaded = true;
+      this.mobilenetModified = net;
+      this.waitingPredictions.forEach(i => this.predict(i.imgToPredict, i.num, i.callback));
+      callback();
+    });
   }
 
-  async predict(input, num, callback) {
-    if (input instanceof HTMLVideoElement && !this.video) {
-      this.video = processVideo(input, '127');
+  async loadModel() {
+    this.mobilenet = await tf.loadModel(this.modelPath);
+    const layer = this.mobilenet.getLayer('conv_pw_13_relu');
+
+    if (this.video) {
+      tf.tidy(() => this.mobilenet.predict(imgToTensor(this.video))); // Warm up
     }
 
-    if (!this.readyPromise) {
-      this.readyPromise = ImageClassifier.loadModel(this.net);
+    return tf.model({ inputs: this.mobilenet.inputs, outputs: layer.output });
+  }
+
+  // Add an image to retrain
+  addImage(labelOrInput, labelOrCallback, cb) {
```

Comment:
I wonder if it makes sense to revisit this naming now that it is really "re-training" the network?

Reply:
Yes! I've been struggling with naming.

Other options were:
addData, addDataClass, updateDataset, AddImageData

Reply:
Yeah, I think I like `addImage()` better than these 😆 so let's keep for now until we think of something else. I try to live by the two words only rule whenever possible.